new blog post: Compare M365 CLI vs SPFx Toolkit vs GitHub Chat Partic… (#2113)
pnp/blog: content/post/different-approaches-in-spfx-development/index.md

---

## 🗒️ Intro

If you are an SPFx developer, you might already be familiar with some of the tooling and approaches that help you out in SPFx development. Most SPFx developers might have started their SPFx development journey with the Yeoman generator, which is the official approach mentioned in Microsoft documentation. That is a totally fine approach, but if you really want to boost your productivity, then you might need to find something more powerful right?

This blog comprehensively explains and boosts your knowledge practically showing how you can use the different toolings available for SPFx development.

I’ll walk-through various approaches for:

- **Setting up your local development environment**
- **Scaffolding project**
- **Managing SPFx solutions in SharePoint Online**

Each approach, whether it's through the command line, the SPFx Toolkit in VS Code, or GitHub Copilot Chat, has its own strengths and trade-offs. Let’s explore them and help you select the right tool for your workflow.

## Setting up your local development environment

Let’s explore in action how each of the tooling options facilitates setting up the local environment for SPFx development.

> ### 1. Microsoft 365 CLI

Reviewer: I know many folks name it Microsoft 365 CLI but the correct name is actually CLI for Microsoft 365. Let's try to use this one in order to avoid confusion 

```suggestion
### 1. CLI for Microsoft 365
```

> With Microsoft 365 CLI, you can manage your Microsoft 365 tenant and SharePoint Framework projects on any platform. No matter if you are on Windows, macOS or Linux, using Bash, Cmder or PowerShell, using the CLI for Microsoft 365 you can configure Microsoft 365, manage SharePoint Framework projects and build automation scripts.

Reviewer: ```suggestion
With CLI for Microsoft 365, you can manage your Microsoft 365 tenant and SharePoint Framework projects on any platform. No matter if you are on Windows, macOS or Linux, using Bash, Cmder or PowerShell, using the CLI for Microsoft 365 you can configure Microsoft 365, manage SharePoint Framework projects and build automation scripts.
```

> Before working with M365 CLI, make sure you have installed it using the following npm script:

Reviewer: ```suggestion
Before working with CLI for Microsoft 365, make sure you have installed it using the following npm script:
```

`npm i -g @pnp/cli-microsoft365`

> After installing it, you need to set it up to start using it. Follow the [M365 CLI documentation](https://pnp.github.io/cli-microsoft365/#getting-started) and complete the setup. Just like what we performed with the other approaches above, M365 CLI offers a command that verifies whether your local environment meets the requirements to work with SharePoint Framework.

Reviewer: ```suggestion
After installing it, you need to set it up to start using it. Follow the [CLI for Microsoft 365 documentation](https://pnp.github.io/cli-microsoft365/#getting-started) and complete the setup. Just like what we performed with the other approaches above, CLI for Microsoft 365 offers a command that verifies whether your local environment meets the requirements to work with SharePoint Framework.
```

![image](images/M365-CLI_1.png)
![image](images/M365-CLI_2.png)
![image](images/M365-CLI_3.png)

This command helps you verify if your environment meets all prerequisites for building solutions using a particular version of the SharePoint Framework.

If you are not inside any SPFx project, the command will try to determine the SharePoint Framework version based on the SharePoint Framework Yeoman generator that you have installed either in the current directory or globally.

### 2. SPFx Toolkit

Lets start with how the local development environment can be set using SPFx Toolkit. To get started with the SharePoint Framework Toolkit, you need to download the extension either from Visual Studio Code itself or from the Visual Studio Marketplace.

a. Action Pane Options

Once it is installed, you will be able to see the SPFx Toolkit icon in the side pane of Visual Studio Code. Clicking on the icon will display the prebuilt actions available in the SPFx Toolkit extension. While we will not delve into each action in detail, let us examine the one most essential for setting up your local environment.

![action pane image](images/SPFx-Toolkit_1.png)

Lets focus on following actions:

> - Check Dependencies: Validates the local setup and checks whether all the required dependencies for working with SPFx are available.

Reviewer: could we align this with the currently latest release?
In the latest minor we just renamed this to `Validate local setup`.
```suggestion
- Validate local setup: Validates the local setup and checks whether all the required dependencies for working with SPFx are available.
```

  If Node.js is not installed, this action will display a message indicating that you need to install the Node.js version required by the latest SPFx specifications. The message also includes links to the supported LTS version of Node.js, and if you prefer using NVM or NVS, relevant links are provided in the dialog as well.

![image comes here](images/SPFx-Toolkit_4.png)

> - Install Dependencies: If the validation using "Check Dependencies" shows ‘Failed validating local setup’, then use this action. It runs an npm installation script to install all the necessary npm modules for SPFx development.

Reviewer: ```suggestion
- Install Dependencies: If the validation using "Validate local setup" shows ‘Failed validating local setup’, then use this action. It runs an npm installation script to install all the necessary npm modules for SPFx development.
```

![img](images/SPFx-Toolkit_7.png)

#### c) SPFx Toolkit - @spfx /new command

As we saw in the previous two capabilities of SPFx Toolkit, let’s now explore what we can do with the @spfx Copilot Chat participant using the `/new` command. Click on the **Copilot** icon to use the `/new` command.

![Image](images/SPFx-Toolkit_9.png)

The `/new` command creates projects by interpreting the natural language you provide and constructing the Yeoman command with the help of AI. In the screenshot above, you can see the AI asking a few questions. Let’s answer those questions and observe the response below.

![Image](images/SPFx-Toolkit_10.png)

In the AI’s response, you’ll see that the LLM constructs a Yeoman script that creates the SPFx component based on your inputs. You can then click **"Create Project"** to scaffold it.

And in case you don’t want to create a new project, you can select the **"Open Sample Gallery"** action, which opens the sample gallery right inside VS Code.

## Manage SPFx Solutions in SharePoint Online

> As we’ve seen how different toolings help developers in setting up the local environment and scaffolding SPFx projects, now let’s explore how these toolings assist in managing SPFx solutions in SharePoint Online. Let’s begin with M365 CLI, and later we’ll dive into the capabilities of the SPFx Toolkit.

Reviewer: ```suggestion
As we’ve seen how different toolings help developers in setting up the local environment and scaffolding SPFx projects, now let’s explore how these toolings assist in managing SPFx solutions in SharePoint Online. Let’s begin with CLI for Microsoft 365, and later we’ll dive into the capabilities of the SPFx Toolkit.
```

> ### M365 CLI

Reviewer: ```suggestion
### CLI for Microsoft 365
```

> For managing SPFx solutions in the App Catalog, M365 CLI offers a wide variety of commands, not only for SPFx apps but also for many other Microsoft 365 operations. Let’s look at a few key commands in detail:

Reviewer: ```suggestion
For managing SPFx solutions in the App Catalog, CLI for Microsoft 365 offers a wide variety of commands, not only for SPFx apps but also for many other Microsoft 365 operations. Let’s look at a few key commands in detail:
```

- spo app add – Allows developers to upload SPFx apps to the tenant-wide App Catalog or site collection level.

> - spo app deploy – After uploading an app to the catalog, this command is used to deploy it. You’ll need to specify the app’s ID or name, along with the App Catalog URL and scope.

Reviewer: ```suggestion
- spo app deploy – After uploading an app to the app catalog, this command is used to deploy it. You’ll need to specify the app’s ID or name, along with the App Catalog URL and scope.
```

> - spo app get – Useful for retrieving metadata of uploaded apps, such as AAD permissions, CDN locations, and more.

Reviewer: AAD is the old name still used under the hood in many variables but we should rather use Entra. 
I think in this case we may just say API permissions or scopes 

```suggestion
- spo app get – Useful for retrieving metadata of uploaded apps, such as Entra permissions, CDN locations, and more.
```

- spo app install – Enables developers to install SPFx apps into specified SharePoint sites.

> 🔗 Check out the full [M365 CLI documentation](https://pnp.github.io/cli-microsoft365/cmd/docs) for more commands

Reviewer: ```suggestion
🔗 Check out the full [CLI for Microsoft 365 documentation](https://pnp.github.io/cli-microsoft365/cmd/docs) for more commands
```

> As we’ve seen, M365 CLI is a powerful tool for managing SPFx apps. But its capabilities go far beyond that, it includes a wide range of commands for managing your entire Microsoft 365 environment.

Reviewer: ```suggestion
As we’ve seen, CLI for Microsoft 365 is a powerful tool for managing SPFx apps. But its capabilities go far beyond that, it includes a wide range of commands for managing your entire Microsoft 365 environment.
```

### SPFx Toolkit Manage Capabilities

> We’ve seen that SPFx Toolkit is a complete package, offering all essential features in one place. Now, let’s explore its potential for managing SPFx apps using the "Manage App Catalog Details" feature and the SPFx Chat Participant with the `/manage` command.

Reviewer: in latest minor we renamed this to `/info` command

```suggestion
We’ve seen that SPFx Toolkit is a complete package, offering all essential features in one place. Now, let’s explore its potential for managing SPFx apps using the "Manage App Catalog Details" feature and the SPFx Chat Participant with the `/info` command.
```

1. Tenant-wide Extensions: Shows a list of extensions that are deployed across the entire tenant.
2. Tenant App catalog: A list of apps, including web parts and extensions, deployed in the tenant for various sites.
3. Site App Catalogs: A list of apps that are deployed under site collection app catalogs is displayed. It also provides the ability to create and remove site collection app catalogs.

- To create new app catalogs at the site level, hover over **Site App Catalogs** and click on the **+ (plus)** icon. Provide the relative URL of the site and confirm to create a new site collection app catalog.
  ![img](images/SPFx-Toolkit_12.png)
  ![img](images/SPFx-Toolkit_11.png)

- Similarly, to remove any site collection app catalog, hover over the app catalogs under **Site App Catalogs** and click on the **delete** icon, which will delete them.
  ![image](images/SPFx-Toolkit_13.png)

When it comes to managing individual apps, SPFx Toolkit provides more flexibility to the developers; you don't need to switch between apps, all operations can be done right from your favorite code editor.

> Just as we saw what M365 CLI offers for managing apps, similarly, SPFx Toolkit offers a similar set of options but in your code editor itself.

Reviewer: ```suggestion
Just as we saw what CLI for Microsoft 365 offers for managing apps, similarly, SPFx Toolkit offers a similar set of options but in your code editor itself.
```
